core/lib/src/response/responder.rs

```diff
@@ -217,34 +220,144 @@ impl<'r> Responder<'r> for String {
+///     fn respond_to(self, req: &Request) -> response::Result<'r> {
+///         Response::build_from(RangeResponder(self).respond_to(req)?)
+///             .header(ContentType::Binary)
+///             .ok()
+///     }
+/// }
+/// ```
+pub struct RangeResponder<B: io::Seek + io::Read>(pub B);
+
+impl<'r, B: io::Seek + io::Read + 'r> Responder<'r> for RangeResponder<B> {
+    fn respond_to(self, req: &Request) -> response::Result<'r> {
+        use http::hyper::header::{ContentRange, ByteRangeSpec, ContentRangeSpec};
+
+        let mut body = self.0;
+        //  A server MUST ignore a Range header field received with a request method other than GET.
+        if req.method() == Method::Get {
+            let range = req.headers().get_one("Range").map(|x| Range::from_str(x));
+            match range {
+                Some(Ok(Range::Bytes(ranges))) => {
+                    if ranges.len() == 1 {
+                        let size = body.seek(io::SeekFrom::End(0))
+                            .expect("Attempted to retrieve size by seeking, but failed.");
+
+                        let (start, end) = match ranges[0] {
+                            ByteRangeSpec::FromTo(mut start, mut end) => {
+                                if start > size {
+                                    start = size;
+                                }
+                                if end > size {
+                                    end = size;
+                                }
+                                (start, end)
+                            },
+                            ByteRangeSpec::AllFrom(mut start) => {
+                                if start > size {
+                                    start = size;
+                                }
+                                (start, size)
+                            },
+                            ByteRangeSpec::Last(len) => {
+                                // we could seek to SeekFrom::End(-len), but if we reach a value < 0, that is an error.
+                                // but the RFC reads:
+                                //      If the selected representation is shorter than the specified
+                                //      suffix-length, the entire representation is used.
+                                let start = size.checked_sub(len).unwrap_or(0);
+                                (start, size)
+                            }
+                        };
+
+                        body.seek(io::SeekFrom::Start(start))
+                            .expect("Attempted to seek to the start of the requested range, but failed.");
+
+                        return Response::build()
+                            .status(Status::PartialContent)
+                            .header(AcceptRanges(vec![RangeUnit::Bytes]))
+                            .header(ContentRange(ContentRangeSpec::Bytes {
+                                range: Some((start, end)),
```

Comment:
At this point `end` is exclusive, so changing to `end-1` converts it to an inclusive position as the client wants.